columnflow/tasks/framework/base.py

```diff
@@ -1213,11 +1194,123 @@ def resolve_param_values(cls, params: dict) -> dict:
-            params = cls.switch_to_analysis_inst_aux(params)
+        # resolving of parameters that is required before ArrayFunctions etc. can be initialized
+        params = cls.resolve_pre_taf_init_params(params)
+
+        # initialize ArrayFunctions etc. and collect known shifts
+        shifts = TaskShifts()
+        params = cls.build_taf_insts(params, shifts)
+
+        # resolving of parameters that can only be performed after ArrayFunction initialization
+        params = cls.resolve_post_taf_init_params(params)
+
+        # resolving of shifts
+        params = cls.resolve_shifts(params, shifts)
+
+        return params
+
+    @classmethod
+    def build_taf_insts(cls, params: dict[str, Any], shifts: TaskShifts) -> dict[str, Any]:
+        """
+        Build the array function instances.
+        For single-config/dataset tasks, build_taf_insts is implemented by mixin classes such as the ProducersMixin.
+        For multi-config tasks, build_taf_insts from the upstream task is called for each config instance.
+        If the build_taf_insts function needs to be called for other combinations of parameters (e.g. per dataset),
+        it can be overwritten by the task class.
+
+        :param params: Dictionary of task parameters.
+        :param shifts: Collection of local and global shifts.
+        :return: Updated dictionary of task parameters.
+        """
+
+        # check if shifts are already known
+        if params.get("known_shifts", None) and params.get("branch", -1) != -1:
+            logger_dev.debug(f"{cls.task_family}: shifts already known")
+            shifts = params["known_shifts"]
+            return params
+        else:
+            if params.get("known_shifts", None) and params.get("branch", -1) == -1:
+                logger_dev.debug(f"{cls.task_family}: shifts already known, but this is branch -1")
+            else:
+                logger_dev.debug(f"{cls.task_family}: shifts unknown")
+            cls.get_known_shifts(params, shifts)
+            params["known_shifts"] = shifts
+
+        cls.get_known_shifts(params, shifts)
+
+        if not cls.upstream_task_cls:
+            params["known_shifts"] = shifts
+            return params
+
+        logger_dev.debug(
+            f"{cls.task_family}: uses ConfigTask.build_taf_insts base implementation; "
+            f"upsteam_task_cls was defined as {cls.upstream_task_cls}; ",
+        )
+        # base implementation for ConfigTasks that do not define any datasets.
+        # Needed for e.g. MergeShiftedHistograms
+        if cls.has_single_config():
+            _params = cls.upstream_task_cls.build_taf_insts(params, shifts)
+            cls.upstream_task_cls.get_known_shifts(params, shifts)
+        else:
+            for config_inst in params["config_insts"]:
+                _params = params.copy()
+                _params["config_inst"] = config_inst
+                _params["config"] = config_inst.name
```

Comment:
Syntax sugar for readability.

```suggestion
                _params = {
                    **params,
                    "config_inst": config_inst,
                    "config": config_inst.name,
                }
```